torch/optim/zeroptimizer.py

```diff
@@ -0,0 +1,524 @@
+        self._per_device_params: OrderedDict[
+            torch.device, List[List[Parameter]]
+        ] = OrderedDict()  # device, rank, params
+        self._param_rank: Dict[torch.Tensor, int] = {}
+        self._partition_parameters: List[List[dict]] = []
+
+        # Build the wrapped optimizer, responsible for a shard of the params
+        self.group = group if group is not None else dist.group.WORLD
+        self.world_size = dist.get_world_size(self.group)
+        self.rank = dist.get_rank(self.group)
+        self.global_rank = _get_global_rank(self.group, self.rank)
+
+        self.optim = optim(self.partition_parameters()[self.rank], **default)
+
+        # - Sync local and global param_groups keys
+        for global_group, local_group in zip(self.param_groups, self.optim.param_groups):
+            for k, v in local_group.items():
+                if k != "params":
+                    global_group[k] = v
+
+        #  Optional consolidated optimizer state
+        self._all_states: List[Dict[str, Any]] = []
+
+        # Current default device is set by the parameters allocated to this rank
+        self._device = self.partition_parameters()[self.rank][0]["params"][0].device
+        self._broadcast_buffers: Dict[torch.device, List[torch.Tensor]] = {}
+        for device, per_device in self.per_device_params.items():
+            # Allocate one buffer per rank and per device to group the small parameters
+            self._broadcast_buffers[device] = [
+                torch.zeros(broadcast_buffer_size, dtype=per_device[0][0].dtype, device=device)
+                for _ in range(len(per_device))
+            ]
+
+    # Partition helpers
+    def partition_parameters(self) -> List[List[dict]]:
+        """Partitions parameters across distributed data parallel ranks.
+
+        Returns a list of param_groups (which is a list of dict) where each
+        element of the list contains the param_groups for a rank. Element 0
+        corresponds to rank 0, etc. We need all the ranks for the broadcast
+        inside step().
+        """
+        if len(self._partition_parameters) == 0:
+            self._partition_parameters = [list() for _ in range(self.world_size)]
+            sizes = [0] * self.world_size
+            for param_group in self.param_groups:
+                param_lists: List[List] = [list() for _ in range(self.world_size)]
+                for param in param_group["params"]:
+                    # Add this param to rank with smallest size.
+                    rank = sizes.index(min(sizes))
+                    param_lists[rank].append(param)
+                    sizes[rank] += param.numel()
+
+                for rank, params in enumerate(param_lists):
+                    param_group_rank = copy.copy(param_group)
+                    param_group_rank["params"] = params
+                    self._partition_parameters[rank].append(param_group_rank)
```

Comment:
I am a little confused here. Looks like `self._partition_parameters` is a `List[Dict]` instead of a `List[List[Dict]]`?

Reply:
it's a 'List[List[Dict]]' as specified